Foundation/NSNotificationQueue.swift

```diff
@@ -15,26 +23,116 @@ public struct NSNotificationCoalescing : OptionSetType {
     
     public init(notificationCenter: NSNotificationCenter) {
-        NSUnimplemented()
+        self.notificationCenter = notificationCenter
+        super.init()
+        NSNotificationQueue.registerQueue(self)
     }
-    
+
+    deinit {
+        NSNotificationQueue.unregisterQueue(self)
+    }
+
     public func enqueueNotification(notification: NSNotification, postingStyle: NSPostingStyle) {
-        NSUnimplemented()
+        enqueueNotification(notification, postingStyle: postingStyle, coalesceMask: [.CoalescingOnName, .CoalescingOnSender], forModes: nil)
     }
 
     public func enqueueNotification(notification: NSNotification, postingStyle: NSPostingStyle, coalesceMask: NSNotificationCoalescing, forModes modes: [String]?) {
-        NSUnimplemented()
+        var runloopModes = [NSDefaultRunLoopMode]
+        if let modes = modes  {
+            runloopModes = modes
+        }
+
+        if !coalesceMask.isEmpty {
+            self.dequeueNotificationsMatching(notification, coalesceMask: coalesceMask)
+        }
+
+        switch postingStyle {
+        case .PostNow:
+            let currentMode = NSRunLoop.currentRunLoop().currentMode
+            if currentMode == nil || runloopModes.contains(currentMode!) {
+                self.notificationCenter.postNotification(notification)
+            }
+        case .PostASAP:
+            self.asapList.append((notification, runloopModes))
+            NSUnimplemented() // post at the end of the current notification callout or timer
+        case .PostWhenIdle:
+            self.idleList.append((notification, runloopModes))
+            NSUnimplemented() // wait until the runloop is idle, then post the notification
+        }
     }
     
-    public func dequeueNotificationsMatching(notification: NSNotification, coalesceMask: Int) {
-        NSUnimplemented()
+    public func dequeueNotificationsMatching(notification: NSNotification, coalesceMask: NSNotificationCoalescing) {
+        var predicate: (NSNotificationListEntry) -> Bool
+        switch coalesceMask {
+        case [.CoalescingOnName, .CoalescingOnSender]:
+            predicate = { (entryNotification, _) in
+                return notification.object === entryNotification.object && notification.name == entryNotification.name
+            }
+        case [.CoalescingOnName]:
+            predicate = { (entryNotification, _) in
+                return notification.name == entryNotification.name
+            }
+        case [.CoalescingOnSender]:
+            predicate = { (entryNotification, _) in
+                return notification.object === entryNotification.object
+            }
+        default:
+            return
+        }
+
+        self.asapList = self.asapList.filter(predicate)
+        self.idleList = self.idleList.filter(predicate)
     }
 
-}
+    // MARK: Private
+
+    private class func registerQueue(notificationQueue: NSNotificationQueue) {
+        var notificationQueueList = self.notificationQueueList
```

Comment:
if this was a NSMutableArray it would encapsulate the hoop jumping of fetching then mutating then re-assigning and the NSMutableArray would act as a box to the array container.